awesomerc.lua

```diff
@@ -439,60 +440,60 @@ end)
 -- {{{ Rules
 -- Rules to apply to new clients.
 -- @DOC_RULES@
-awful.rules.rules = {
+ruled.client.connect_signal("request::rules", function()
     -- @DOC_GLOBAL_RULE@
     -- All clients will match this rule.
-    { rule = { },
-      properties = { focus = awful.client.focus.filter,
-                     raise = true,
-                     screen = awful.screen.preferred,
-                     placement = awful.placement.no_overlap+awful.placement.no_offscreen
-     }
-    },
+    ruled.client.append_rule {
```

Comment:
> That seems pointless.

I could invent some convoluted reasons, but there is actually a good one for this. First of all, doing this for the sake of consistency is good enough on its own.

However the real reason is the new screen API (and its rules, coming after this PR gets merged). I want to get rid of all user code running in the global context (code directly in the top-level indentation rather than a signal handler) and eventually fully deprecate doing so. With the screen rules, `mouse.screen` is `nil` when AwesomeWM starts, which is unsupported by most APIs. The theme API is also uninitialized, which is already causing issues if people run code ahead of `beautiful.init()`. On top of that, if users run code before `awful` is `require()`d, it also already blows up quite bad.

Updating the documentation to say what is and isn't available would to impossible to maintain and make the user experience more complicated that is has to be.

All in all, running code in the global context is already too fragile to support. Having a "managed" initialization where everything is called only after it becomes "really" available is much safer and easier to support.

> rules can be added to instead of having to call some magic global function

There is such an object (`gears.matcher`), but its hidden. Using that as a signal parameter would be interesting. It would allow to deprecate the global API. The magic `::connected` signal API could be used to make sure all new `connect_signal` gets called even if they come later.

The main problem with this would be that the full `gears.matcher` API cannot currently get exposed for all rules type. I fear that exposing such low-level object would enable some self-foot-shooting opportunities comparing to restricting the API to a known safe subset. Given that each type of rules (client, notification, screen, tags, clipboard) have a different safe subset, I am not sure how this could be documented. Given only 1 `gears.matcher`. object can exist per type of rules, I don't see why using the object oriented approach here would make the API simpler.

In any case, if you think we should still use objects in the signal callback, I will do it. I am fine either way.

> (which I think they should NOT do; what's a valid use case for that?!?)

Mostly simplicity. The default config has some built-in rules. A module could use the `id` to disable them. This is simpler for the non-programmer users than having instructions to remove code from `rc.lua`. That's about the only "good" use case.

The reason there is an `append_`/`remove_` is to allow modules to add new rules without fear that another one (or `rc.lua`) will just set `awful.rules.rules` directly and overwrite what they just did. Remove is added to all those APIs for consistency rather than having a real use case.